File: AutoVE_actions/actions.py
import subprocess
import os

def convert_video_format(source_path, target_path):
    """
    Convert the video format from one to another.

    Args:
    source_path : The path to the source video file.
    target_path : The path to save the converted video file.
    """
    print(f"Converting video format from {source_path} to {target_path}")
    subprocess.run(["ffmpeg", "-i", source_path, "-acodec", "copy", "-vcodec", "copy", target_path], check=True)

def extract_audio_from_video(source_path, target_path):
    """
    Extract the audio from a video file.

    Args:
    source_path : The path to the source video file.
    target_path : The path to save the extracted audio file.
    """
    print("Extracting audio from video")
    subprocess.run(["ffmpeg", "-i", source_path, "-q:a", "0", "-map", "a", target_path], check=True)

def resize_video(source_path, width, height, target_path):
    """
    Resize the video to the specified dimensions.

    Args:
    source_path : The path to the source video file.
    width : The width of the resized video.
    height : The height of the resized video.
    target_path : The path to save the resized video file.
    """
    scale = f"scale={width}:{height}"
    print(f"Resizing video to {width}x{height}")
    subprocess.run(["ffmpeg", "-i", source_path, "-vf", scale, target_path], check=True)

def trim_video(source_path, start_time, end_time, target_path):
    """
    Trim the video to the specified start and end times.

    Args:
    source_path : The path to the source video file.
    start_time : The start time of the trimmed video.
    end_time : The end time of the trimmed video.
    target_path : The path to save the trimmed video file.
    """
    print(f"Trimming video from {start_time} to {end_time}")
    subprocess.run(["ffmpeg", "-i", source_path, "-ss", start_time, "-to", end_time, "-c", "copy", target_path], check=True)


def open_video(video_path):
    try:
        # Normalize the path to ensure it's in the correct format for the OS
        normalized_path = os.path.normpath(video_path)
        
        # Ensure the path is correctly quoted to handle spaces and special characters
        command = ['ffplay', '-autoexit', f'"{normalized_path}"']
        
        # Use shell=True on Windows to handle path correctly
        subprocess.run(' '.join(command), check=True, shell=True)
    except subprocess.CalledProcessError:
        print("Failed to open video.")

File: AutoVE/UtilsLib/Actions.cpp
/**
 * @file Actions.cpp
 * @author srira
 */

#include "Actions.h"
#include <iostream>


void convert_video_format(const std::string& source_path, const std::string& target_path) {
	std::string command = "ffmpeg -i " + source_path + " -acodec copy -vcodec copy " + target_path;
	std::cout << "Converting video format from " << source_path << " to " << target_path << std::endl;
	int result = std::system(command.c_str());
	if (result != 0) {
		std::cerr << "Error: ffmpeg command failed with exit code " << result << std::endl;
	}
}

void extract_audio_from_video(const std::string& source_path, const std::string& target_path) {
	std::string command = "ffmpeg -i " + source_path + " -q:a 0 -map a " + target_path;
	std::cout << "Extracting audio from video" << std::endl;
	int result = std::system(command.c_str());
	if (result != 0) {
		std::cerr << "Error: ffmpeg command failed with exit code " << result << std::endl;
	}
}

void resize_video(const std::string& source_path, int width, int height, const std::string& target_path) {
	std::string scale = "scale=" + std::to_string(width) + ":" + std::to_string(height);
	std::string command = "ffmpeg -i " + source_path + " -vf " + scale + " " + target_path;
	std::cout << "Resizing video to " << width << "x" << height << std::endl;
	int result = std::system(command.c_str());
	if (result != 0) {
		std::cerr << "Error: ffmpeg command failed with exit code " << result << std::endl;
	}
}

void trim_video(const std::string& source_path, const std::string& start_time, const std::string& end_time, const std::string& target_path) {
	std::string command = "ffmpeg -i " + source_path + " -ss " + start_time + " -to " + end_time + " -c copy " + target_path;
	std::cout << "Trimming video from " << start_time << " to " << end_time << std::endl;
	int result = std::system(command.c_str());
	if (result != 0) {
		std::cerr << "Error: ffmpeg command failed with exit code " << result << std::endl;
	}
}

//int main() {
//	std::string source_path = "input.mp4";
//	std::string target_path_video = "output.mp4";
//	std::string target_path_audio = "output.mp3";
//	std::string target_path_resized = "resized.mp4";
//	std::string target_path_trimmed = "trimmed.mp4";
//
//	convert_video_format(source_path, target_path_video);
//	extract_audio_from_video(source_path, target_path_audio);
//	resize_video(source_path, 1280, 720, target_path_resized);
//	trim_video(source_path, "00:00:10", "00:00:20", target_path_trimmed);
//
//	return 0;
//}